### ISSUE TITLE
"Pinned Branches" Section

### ISSUE THREAD

**Issue by clounie** [ForkIssues/Tracker#15]
Going off the brief discussion in #14, it would be nice to have a pinned branches section for frequently used branches, or those you're currently focusing on.

I'll generally have a set of branches I'm working on per sprint, and it would be awesome to have those in a group above the normal branch listing.

**Comment by DanPristupov** [comment]
Hi. I'll figure out how to add this (from the UX perspective). May be there's a place for one more icon near the funnel button.

Just brainstorming, can a "recent branches" section somewhere solve this issue?

**Comment by cdzombak** [comment]
"Recent Branches" would mostly solve this for me, I think.

For perspective, my two use cases:

- Most of our active development takes place on the "develop" branch. That's actually set as the default branch on GitHub. 

- I typically am working on a few feature branches at any given time.

It'd be good to surface both of those in the branches list somehow.

**Comment by clounie** [comment]
I think trying to force it into Recent Branches might complicate things more than is necessary.

How do you mesure "Recent"?  
- Your last commit?
- Your last push?
- The last commit by anyone? 

How many recent branches do you display? The user should be able to configure that number if they can't control which ones are categorized there.

Personally would much rather just be able to select the branches I want - but if you were able to specify those two settings (definition of recent + number of branches) then that does solve the problem.

**Comment by cdzombak** [comment]
> How do you mesure "Recent"?

GitHub's "branches" UI solves this by listing the default branch, "your" branches, "recent" branches from everybody, and "stale" branches.

This is more complexity than can reasonably be presented in a folded list UI, IMO. Still, I'd welcome _some_ way to surface the branches I listed in https://github.com/ForkIssues/Tracker/issues/15#issuecomment-297716309.

**Comment by tangobravo** [comment]
Nice work on Fork! I'm basically looking for a client that lets me do multi-funneling #73 so unfortunately Fork is not yet the answer for me. But as you seem to be actively working on the project and responding to feedback thought I'd throw in some ideas here.

The panel on top of the sidebar appears to control the view in the main pane. So how about a "+" in that section to allow you to add "New Commit Filter". That would allow you to enter a name such as "My Branch Set", which would then appear below the "All Commits" entry.

When one of the named commit filters is selected in the top sidebar section, a new column of checkboxes would appear in the lower sidebar panel next to each of the refs (local and remote branches, tags, and stashes). The ones that are ticked would be the filters applied on the main history view; as if they were multi-funnelled as requested in #73. Could be a column of funnel icons too, but I'd expect them to be shown for each row if one of the named "Commit Filter" entries has been selected rather than individually appearing on mouse-over.

The advantage of this approach is that you can have multiple named sets of filtered branches that you can switch between. It also feels like a better match to the repo itself to display them as filters on the full ref list in the lower panel (rather than adding another "pinned branches" section in that lower panel, and duplicating branch names) - the top sidebar panel feels more like Fork-app specific options, and the lower panel is the true repo state as seen by git.

Additionally a toggle could be added in the ref filter box (currently at the bottom of the sidebar), so only the selected rows would be shown. I'd also vote for moving the filter to the top of that lower panel, and having the state of that toggle saved along with the set of selected branches.

And sorry for throwing so many ideas in a single comment; but additionally "Local Branches Only" and "Local and Tracking Branches Only" views could nicely fit in as default options underneath that "All Commits" entry.

**Comment by DanPristupov** [comment]
@tangobravo Your post a bit overwhelmed me. Awesome ideas 👍. I need to get a pause and think few days and then get back to you.

**Comment by DanPristupov** [comment]
After devoting some time to thinking about it I came to the conclusion that named sets of filtered branches sound exciting but probably won't be very useful in the real life.

I just can't imagine a real usage example.

- Group like `develop, master` doesn't make much sense, because we rarely need to have them together. Much more useful will be an ability to pin these branches.
- Group like `milestone1/feature1, milestone1/feature2` features for a milestone. User must keep them up to date, otherwise such groups will quickly become a mess. People are used to create and merge branches a lot, all such groups will quickly become absolete.

What are yours use cases?

I think having pinned branches (stars) and multi-funneling will solve 95% of requirements. (BTW, I'm planning to implement both these features in the next release.)

P.S. Branch filter groups remind me of the ability to save searches in Finder. It's seems to be a very powerful feature, but I've never used it. I just don't know how to fit it in my working process.

**Comment by tangobravo** [comment]
Fair enough. Although there's obviously common practice around branches, eg features and milestones, at the low level a git repo is just the DAG of commits with various references to individual commits in that DAG. It's quite possibly my fault that my repos get into a bit of a mess and end up hard to follow with too many branches.

When trying to wrap my head around the history, a single-branch filter just isn't enough, but multi-funnelling would get most of the way there for me (UX-wise just a cmd-click in the funnel column would work fine I'd suggest).

Assigning a name to that multi-funnelled set seems like a potentially useful small extension on that; but I absolutely understand the benefit in keeping the UX clean and uncluttered and its your call to make.

Can you say a bit more about how you envisage the UX for Pinned Branches working? In what way is the concept distinct from a saved multi-funnel set? In this ticket at least it's not clear exactly what is proposed.

Also happy to sit tight and wait and see what you come up with in the next version!

**Comment by nebhale** [comment]
An example might be repository, used by a large-ish team, that contains both multiple release branches (`master`, `2.x`, `2.0.x`, `1.x`, etc.) as well a number of in-flight development branches (`409-cleanup`, `421-resource-matching`, `455-support-cfignore`, etc.) pushed by various team members.  Most of the time, a given team member only cares about the release branches and their current development branch; the rest of the development branches are “noise” on the railroad diagram that clutter their understanding of the single development branch.  Especially when it comes time to forward merge a fix across multiple development branches (fix on the outside most branch and bring forward to `master`, fixing all supported versions along the way) having a clean idea of the state of the “main” branches is a win.  The same kind of thinking applies when trying to find where a recent commit (which may be out on an outer branch somewhere) causes a breakage; reducing the complexity of the diagram to a smaller set of branches shrinks the problem set and makes it easier to find.

My preferred behavior would be that when viewing a repository, any pinned branches (or branch groups) _as well as the currently checked out branch_ would be visible and all others hidden.  My workflow would likely contain a single pinned group of the release branches, and that along with my current development branch being visible would allow me to focus the change I made and how it would affect all of the versions that we’re currently supporting.  It would allow me to ignore other development currently in-flight from my team.

**Comment by DanPristupov** [comment]
Guys, I released Fork 1.0.56 update which has an ability to pin branches (both local and remote) and tags. The option is available in context menu.

![2017-09-20 at 21 08](https://user-images.githubusercontent.com/618115/30662611-f1796664-9e47-11e7-9eac-044a59b86524.gif)
![2017-09-20 at 21 17](https://user-images.githubusercontent.com/618115/30662928-19984330-9e49-11e7-9836-981bfac65232.png)

Honestly, I feel that something should be improved there, but I'm not sure what. May be I should make pinned branches more contrast and draw some stars near them.

Any feedback and comments are welcome.

P.S. I didn't complete multi-funnel yet, because I rushed to release update to solve problems with Xcode9 update.

**Comment by clounie** [comment]
Been using it for a day now, and loving it. 

I hadn't really thought about pinning remote branches, but having a separate section for those is really nice (pull latest changes from dev into your branch? So much easier, don't have to chase it down among 100s of branches.)

Only small issue I found - it seems like the triangle under remotes likes to close itself. 

For example, if my state is like this:
![image](https://user-images.githubusercontent.com/11947182/30760944-a55f31de-9f99-11e7-889d-d611de2a4008.png)
..and I close Fork, then I get this upon re-opening the app:
![image](https://user-images.githubusercontent.com/11947182/30760959-b2445244-9f99-11e7-8f2e-c87960c34b3e.png)

It would be preferable for this to maintain its state prior to close - the normal "origin" under remotes does this properly already.

**Comment by DanPristupov** [comment]
@clounie thank you for the feedback and for the bug report.

Is context menu a good place for that option? Would it be better if I show a star on hovering a row (similar to funnel icon)?

**Comment by clounie** [comment]
I don't think I explained my issue very well.

My problem is that the remotes under Quick Access do not maintain their expanded/collapsed state in the same manner remotes under the Remotes section.

### Steps to Reproduce:
1. Pin a branch belonging to a remote (say, `origin`)
2. Click the triangle to expand that `origin` remote and show branches underneath it (this remote is under the *Quick Access* section)
3. Now, expand one of your remotes in the *Remotes* section
4. Close app and re-open

### Problem: 
When you re-open, the remote from step 3 will have stayed expanded, but the remote in the *Quick Access* section from step 2 will have collapsed.

### Desired behavior:
**Given:** There is a branch in the *Quick Access* section which belongs to a remote
+
**Given:** I have expanded the remote

**When:** I close + re-open Fork
**Then:** I expect that remote in the *Quick Access* section to have maintained its expanded state from before I closed the app

**Comment by DanPristupov** [comment]
@tangobravo @nebhale @clounie Guys, I implemented multi-funneling and it's available in 1.0.59.4 prerelease (https://github.com/ForkIssues/Tracker/issues/152)!

**Comment by clounie** [comment]
The star is great!

Still wish that the `Starred/origin` folder would preserve its open/closed state when I close and re-open Fork - notice that every time it happens.

**Comment by clounie** [comment]
Also, see the following:
![image](https://user-images.githubusercontent.com/11947182/32920065-3332d152-cae5-11e7-9808-f53bcb14eaf2.png)

The starred tooltip appears even when you hover on the hourglass.

**Comment by DanPristupov** [comment]
@clounie the bug with open/closed state can't be fixed quickly. It's still in my todo list.
Regarding the tooltip: will fix it in the next update.

Thank you
